### PR TITLE
dynamic fee txs (EIP-1559 support)

### PR DESCRIPTION
**Pull request by MrChico**
No algorithms for setting prio fee, just a good old manual flags allowing for eip1559 txs. 

Haven't tried this out much yet, brave guinea pigs appreciated

### CODE REVIEW

**Comment by MrChico** [comment]
Actually, maybe better to not distinguish `max-fee` and `gas-price`? And treat txs as dynamic fee txs when `prio-fee` is set

File: src/seth/libexec/seth/seth-mktx [hunk 1]
```diff
@@ -29,6 +28,18 @@ else
   args+=(--to "$(seth --to-address "$1")")
 fi
```
Comment (teghnet) [inline src/seth/libexec/seth/seth-mktx:30]:
```suggestion

if [[ $ETH_GAS_PRICE && ($ETH_PRIO_FEE || $ETH_MAX_FEE) ]]; then
  echo >&2 "required: either only the gas price or both max prio and max fee"         
  seth --fail-usage "$0"
fi

```

File: src/seth/libexec/seth/seth [hunk 2]
```diff
@@ -153,6 +156,8 @@ while [[ $1 ]]; do
     -G|--gas)        shift; export ETH_GAS=$1;;
     -P|--gas-price)  shift;
       export ETH_GAS_PRICE=$(seth --to-wei ${1%gwei} ${1##*([0-9])});;
+    --prio-fee)      shift;
```
Comment (teghnet) [inline src/seth/libexec/seth/seth:159]: is the `gas price` going to be used as `max fee` when the tip is set?

File: src/seth/libexec/seth/seth-mktx [hunk 3]
```diff
@@ -29,6 +29,15 @@ else
...
 
+if [[ $ETH_PRIO_FEE ]]; then
+  args+=(--prio-fee "$ETH_PRIO_FEE")
+else
+  echo >&2 "required: gas price or (max prio and max fee)"
```
Comment (teghnet) [inline src/seth/libexec/seth/seth-mktx:35]: the error message seems inaccurate

**Review by teghnet — COMMENTED** [review]
Hi, I really like that you started working on adding EIP-1559 tx params to `seth`.
Do you mind if I comment on this?

**Comment by teghnet** [comment]
> Actually, maybe better to not distinguish `max-fee` and `gas-price`? And treat txs as dynamic fee txs when `prio-fee` is set

I agree with adding only `tip` option. Less options less problems.
My reasoning is that gas price is already being used as max fee, and miner tip is set to the same value already by ethereum, so if we just update the latter, we're fine and everyone knows what is what - we are consistent with node's logic.

File: src/seth/libexec/seth/seth [hunk 4]
```diff
@@ -71,6 +72,8 @@ keystore=dir         local keystore directory
 F,from=address       account to send transaction from
 G,gas=number         number of gas units to provide
 P,gas-price=number   price (in wei) of each gas unit
+max-prio=number      max priority fee (miner tip) in wei per gas
```
[the remark this answers is not in the document]
Comment (MrChico) [reply src/seth/libexec/seth/seth:75]: Good idea!
```suggestion
max-prio=number    max priority fee (miner tip) in wei per gas
```

File: src/ethsign/ethsign.go [hunk 5]
```diff
@@ -349,7 +354,7 @@ func main() {
       },
       Action: func(c *cli.Context) error {
         requireds := []string{
-          "nonce", "value", "gas-price", "gas-limit", "chain-id", "from",
+          "nonce", "value", "gas-limit", "gas-price", "chain-id", "from",
```
Comment (MrChico) [inline src/ethsign/ethsign.go:357]:
```suggestion
          "nonce", "value", "gas-price", "gas-limit", "chain-id", "from",
```

File: src/seth/CHANGELOG.md [hunk 6]
```diff
@@ -10,6 +10,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 - seth bundle-source writes the contents of standard-json to the current directory to enable better sourcemaps for multi-file etherscan source code.
 
+### Added 
+
+- Dynamic transaction fee format transactions (EIP-1559) supported by introducing the flag `--prio-fee` and corresponding environment variable `ETH_PRIO_FEE` to seth and ethsign. If `--prio-fee` is provided (or `ETH_PRIO_FEE`) is set, 1559 transaction will be used and `--gas-price` will reflect the maximum gas price, rather than the absolute gas price.
+
+### Changed
+
+- `--gas-price` must be explicitly set: seth will no longer ask the rpc client for appriopriate pricing.
+
```
Comment (MrChico) [inline src/seth/CHANGELOG.md:20]:
```suggestion
### Changed

- `--gas-price` will be used as `max-fee` when `--prio-fee` is set

```

File: src/seth/libexec/seth/seth [hunk 7]
```diff
@@ -19,6 +19,7 @@
 ###    -F, --from=<sender>        account to send transaction from
 ###    -G, --gas=<number>         number of gas units to provide
 ###    -P, --gas-price=<number>   price (in wei) of each gas unit
+###    --max-prio=<number>        max priority fee (miner tip) in wei per gas
```
Comment (MrChico) [inline src/seth/libexec/seth/seth:22]:
```suggestion
###    -T, --max-prio=<number>    max priority fee (miner tip) in wei per gas
```

File: src/seth/libexec/seth/seth (hunk 2 above)
Comment (MrChico) [reply src/seth/libexec/seth/seth:159]: yes

Comment (MrChico) [reply src/seth/libexec/seth/seth:159]:
```suggestion
    -T|--prio-fee)   shift;
```

File: src/seth/libexec/seth/seth-mktx (hunk 3 above)
Comment (MrChico) [reply src/seth/libexec/seth/seth-mktx:35]: yup, thanks

File: src/seth/libexec/seth/seth-mktx [hunk 8]
```diff
@@ -29,6 +29,15 @@ else
   args+=(--to "$(seth --to-address "$1")")
 fi
 
+if [[ $ETH_PRIO_FEE ]]; then
+  args+=(--prio-fee "$ETH_PRIO_FEE")
+else
+  echo >&2 "required: gas price or (max prio and max fee)"         
+  seth --fail-usage "$0"
+fi
```
Comment (MrChico) [inline src/seth/libexec/seth/seth-mktx:37]:
```suggestion
fi
```

**Comment by MrChico** [comment]
Thanks for the comments @teghnet I appreciate it

File: src/seth/libexec/seth/seth [hunk 9]
```diff
@@ -19,6 +19,7 @@
 ###    -F, --from=<sender>        account to send transaction from
 ###    -G, --gas=<number>         number of gas units to provide
 ###    -P, --gas-price=<number>   price (in wei) of each gas unit
+###    -T, --max-prio=<number>    max priority fee (miner tip) in wei per gas
```
Comment (MrChico) [inline src/seth/libexec/seth/seth:22]:
```suggestion
###    -T, --prio-fee    max priority fee (miner tip) in wei per gas
```

File: src/seth/libexec/seth/seth [hunk 10]
```diff
@@ -71,6 +72,8 @@ keystore=dir         local keystore directory
 F,from=address       account to send transaction from
 G,gas=number         number of gas units to provide
 P,gas-price=number   price (in wei) of each gas unit
+T,max-prio=number    max priority fee (miner tip) in wei per gas
```
Comment (MrChico) [inline src/seth/libexec/seth/seth:75]:
```suggestion
T,prio-fee=number    max priority fee (miner tip) in wei per gas
```

File: src/seth/README.md [hunk 11]
```diff
@@ -703,17 +703,19 @@ Sign and publish a transaction to the blockchain.
...
-| `--resend`    | `SETH_RESEND`   |              | reuse nonce     |
-| `--async`     | `SETH_ASYNC`    |              | don't wait      |
-| `--status`    | `SETH_STATUS`   |              | check success   |
+| Flag          | Variable        | Default      | Synopsis                          |
+| ------------- | --------------- | ------------ | ---------------                   |
+| `--block`     | `ETH_BLOCK`     | `latest`     | block number                      |
+| `--from`      | `ETH_FROM`      | n/a          | sender                            |
+| `--gas`       | `ETH_GAS`       | node decides | gas quantity                      |
+| `--gas-price` | `ETH_GAS_PRICE` |              | gas price                         |
+| `--max-fee`   | `ETH_MAX_FEE`   |              | max net gas cost                  |
```
Comment (MrChico) [inline src/seth/README.md:712]:
```suggestion
```

File: src/seth/README.md [hunk 12]
```diff
@@ -726,6 +727,9 @@ the transaction receipt and exit with an error code if the transaction
 failed. This is a post-Byzantium feature and will soon become the
 default behavior.
 
+If `--gas-price` is provided (or `ETH_GAS_PRICE`) is set, legacy transactions will be used.
+For dynamic fee transactions (EIP-1559), `--max-fee` and `--prio-fee` is required.
```
Comment (MrChico) [inline src/seth/README.md:731]:
```suggestion
For dynamic fee transactions (EIP-1559), `--prio-fee` is required.
```

File: src/dapp-tests/integration/tests.sh [hunk 13]
```diff
@@ -46,6 +46,9 @@ dapp_testnet() {
 
   # since we have one tx per block, seth run-tx and seth debug are equivalent
   [[ $(seth run-tx "$TX") = 0x ]] || error
+
+  # dynamic fee transaction (EIP-1559)
+  seth send "$A_ADDR" "off()" --gas 0xffff --password /dev/null --from "$ACC" --keystore "$TMPDIR"/8545/keystore --prio-fee 2 'gwei' --gas-price 10 'gwei' --async
```
Comment (MrChico) [inline src/dapp-tests/integration/tests.sh:51]:
```suggestion
  seth send "$A_ADDR" "off()" --gas 0xffff --password /dev/null --from "$ACC" --keystore "$TMPDIR"/8545/keystore --prio-fee 2gwei --gas-price 10gwei
```

**Comment by MrChico** [comment]
no idea why this is failing now

**Comment by endophysics** [comment]
I did a bit of digging around and it seems that in `ethsigner.go` instead of

```Go
encoded, _ := rlp.EncodeToBytes(signed)
```

It has to be

```Go
encoded, _ := signed.MarshalBinary()
```

The first one will produce an RLP encoded string such as 
```
0xb86302f8608001020a0a94ea76c3bc8c659391464c68547255cfa955b19ea48080c001a03b181afcd28b3ce4bd1e86b064e6333b812707039bfe3d777d5105ef80e717bba050d7b79cc370c228f9374df9eab5d7c1089de3331b1c9f0156f08d7476235923
```

but when that is fed into `eth_sendRawTransaction` [here](https://github.com/ethereum/go-ethereum/blob/f38abc55f135587ce3cf60c9d574ec2a4ebb8197/internal/ethapi/api.go#L1761) it expects the binary version, so if you look at [`UnmarshalBinary`](https://github.com/ethereum/go-ethereum/blob/f38abc55f135587ce3cf60c9d574ec2a4ebb8197/core/types/transaction.go#L155) it'll check if if the first byte is `> 0x7f` as per EIP2718, will find `0xb8` and assume it's a legacy tx, but then not find a list encoding.

**Comment by MrChico** [comment]
Thanks @endophysics, with that change I was able to get our test passing. This is now be ready for review and more testing

File: src/seth/libexec/seth/seth [hunk 14]
```diff
@@ -19,6 +19,7 @@
 ###    -F, --from=<sender>        account to send transaction from
 ###    -G, --gas=<number>         number of gas units to provide
 ###    -P, --gas-price=<number>   price (in wei) of each gas unit
+###    -T, --prio-fee    max priority fee (miner tip) in wei per gas
```
Comment (d-xo) [inline src/seth/libexec/seth/seth:22]: nit: alignment

File: src/seth/libexec/seth/seth [hunk 15]
```diff
@@ -71,6 +72,8 @@ keystore=dir         local keystore directory
 F,from=address       account to send transaction from
 G,gas=number         number of gas units to provide
 P,gas-price=number   price (in wei) of each gas unit
+T,prio-fee=number    max priority fee (miner tip) in wei per gas
+
```
Comment (MrChico) [inline src/seth/libexec/seth/seth:76]:
```suggestion
```

File: src/seth/libexec/seth/seth-mktx [hunk 16]
```diff
@@ -29,6 +29,12 @@ else
   args+=(--to "$(seth --to-address "$1")")
 fi
 
+if [[ $ETH_PRIO_FEE ]]; then
+  args+=(--prio-fee "$ETH_PRIO_FEE")
+fi
+
+
+
```
Comment (d-xo) [inline src/seth/libexec/seth/seth-mktx:37]:
```suggestion
```

File: src/dapp-tests/integration/tests.sh [hunk 17]
```diff
@@ -46,6 +46,9 @@ dapp_testnet() {
 
   # since we have one tx per block, seth run-tx and seth debug are equivalent
   [[ $(seth run-tx "$TX") = 0x ]] || error
+
+  # dynamic fee transaction (EIP-1559)
+  seth send "$A_ADDR" "on()" --gas 0xffff --password /dev/null --from "$ACC" --keystore "$TMPDIR"/8545/keystore --prio-fee 2gwei --gas-price 10gwei
```
Comment (d-xo) [inline src/dapp-tests/integration/tests.sh:51]: could we expand the test a little to check that the correct gas price is used?

File: src/seth/libexec/seth/seth (hunk 14 above)
Comment (MrChico) [reply src/seth/libexec/seth/seth:22]:
```suggestion
###    -T, --prio-fee               max priority fee (miner tip) in wei per gas
```

File: src/seth/libexec/seth/seth [hunk 18]
```diff
@@ -19,6 +19,7 @@
 ###    -F, --from=<sender>        account to send transaction from
 ###    -G, --gas=<number>         number of gas units to provide
 ###    -P, --gas-price=<number>   price (in wei) of each gas unit
+###    -T, --prio-fee               max priority fee (miner tip) in wei per gas
```
Comment (MrChico) [inline src/seth/libexec/seth/seth:22]:
```suggestion
###    -T, --prio-fee            max priority fee (miner tip) in wei per gas
```

File: src/seth/libexec/seth/seth [hunk 19]
```diff
@@ -19,6 +19,7 @@
 ###    -F, --from=<sender>        account to send transaction from
 ###    -G, --gas=<number>         number of gas units to provide
 ###    -P, --gas-price=<number>   price (in wei) of each gas unit
+###    -T, --prio-fee            max priority fee (miner tip) in wei per gas
```
Comment (MrChico) [inline src/seth/libexec/seth/seth:22]:
```suggestion
###    -T, --prio-fee             max priority fee (miner tip) in wei per gas
```

**Comment by teghnet** [comment]
Can we merge it now?

**Review by d-xo — APPROVED** [review]

**Pull request merged**